rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp

```diff
@@ -162,17 +191,32 @@ void SequentialWriter::remove_topic(const rosbag2_storage::TopicMetadata & topic
   }
 }
 
+bool SequentialWriter::compress_file_and_update_metadata()
+{
+  if (compressor_ && compression_mode_ == CompressionMode::FILE) {
+    // Get the uri of the last file and pop the uri
+    const auto uncompressed_uri = metadata_.relative_file_paths.back();
```

Comment:
Not sure how you guarantee `metadata_.relative_file_paths.back()` isn't already an URI to a compressed file. For example, `split_bagfile()` and `reset()` are called right after the other.

Reply:
Maybe a helper function to check if the URI is already compressed (The extension matches the compression identifier)?

Reply:
The happy case seems to be: `split_bagfile()` is called, the next rosbag starts being recorded a new element is pushed back on `metadata_.relative_file_paths`, then recording is stopped and `reset()` is called.

The bad case is: `split_bagfile()` is called, then recording is stopped and `reset()` is called. This still feels like a pretty normal situation.

It's probably better to have a flag that tracks whether `metadata_.relative_file_paths.back()` has been compressed already or not, which gets set when a new rosbag starts and a when the current rosbag ends (I can't tell if this needs to be also threadsafe or not). Just checking the extension doesn't scale when you have more compression algorithms / identifiers available.

Reply:
I added a flag `should_compress_last_file_` but maybe all we need to do is check if we are splitting or not (since we know this is a situation that only occurs when splitting).
This just means checking for `max_bagfile_size_ == 0`.

Reply:
> However, `split_bagfile()` always adds a new file to the end so eventually, it should always be compressed on `reset()`.

Sorry, I missed the `metadata_.relative_file_paths.push_back()` in `split_bagfile()`, so I thought it was happening elsewhere.

However, I also noticed in `split_bagfile()` that you're potentially throwing between `compress_last_file()` and `metadata_.relative_file_paths.push_back()`, so you can still potentially end up with `metadata_.relative_file_paths.back()` being a file that's already compressed.

I'm not sure how such a thrown exception will be handled. How do you guarantee that the next time you enter `split_bagfile()`, `metadata_.relative_file_paths.back()` is not already compressed?

Reply:
> However, I also noticed in `split_bagfile()` that you're potentially throwing between `compress_last_file()` and `metadata_.relative_file_paths.push_back()`, so you can still potentially end up with `metadata_.relative_file_paths.back()` being a file that's already compressed.

That's correct, it seems a `bool` flag for that small section that could throw is required to make sure we don't double compress.